extractor/filesystem/language/javascript/packagelockjson/extractor.go

```diff
@@ -1,37 +1,236 @@
-// Copyright 2024 Google LLC
-//
-// Licensed under the Apache License, Version 2.0 (the "License");
-// you may not use this file except in compliance with the License.
-// You may obtain a copy of the License at
-//
-//      http://www.apache.org/licenses/LICENSE-2.0
-//
-// Unless required by applicable law or agreed to in writing, software
-// distributed under the License is distributed on an "AS IS" BASIS,
-// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-// See the License for the specific language governing permissions and
-// limitations under the License.
-
-// Package packagelockjson extracts package-lock.json files.
+// Package packagelockjson extracts yarn.lock files.
 package packagelockjson
 
 import (
 	"context"
+	"encoding/json"
 	"fmt"
 	"io/fs"
+	"path"
 	"path/filepath"
-	"slices"
 	"strings"
 
-	"github.com/google/osv-scanner/pkg/lockfile"
 	"github.com/google/osv-scalibr/extractor"
 	"github.com/google/osv-scalibr/extractor/filesystem"
+	"github.com/google/osv-scalibr/extractor/filesystem/language/javascript/internal/commitextractor"
 	"github.com/google/osv-scalibr/extractor/filesystem/osv"
 	"github.com/google/osv-scalibr/plugin"
 	"github.com/google/osv-scalibr/purl"
 	"github.com/google/osv-scalibr/stats"
+
+	"golang.org/x/exp/maps"
+	"golang.org/x/exp/slices"
 )
 
+type npmLockDependency struct {
+	// For an aliased package, Version is like "npm:[name]@[version]"
+	Version      string                       `json:"version"`
+	Dependencies map[string]npmLockDependency `json:"dependencies,omitempty"`
+
+	Dev      bool `json:"dev,omitempty"`
+	Optional bool `json:"optional,omitempty"`
+
+	Requires map[string]string `json:"requires,omitempty"`
```

Comment:
A bunch of these doesn't seem to be used (Requires, DevDependencies, OptionalDependencies, etc.). I'd say we should remove them from the json struct if we don't actually use  them anywhere.

Reply:
They are used in depGroups() functions to populate the depgroups metadata.

Reply:
Yeah I'd say let's add it back when GR gets migrated.